<launch>

  <arg name="robot" default="firefly_sbx"/>
  <arg name="fixed_frame_id" default="world"/>
  <arg name="calibration_file" default="$(find msckf_vio)/config/camchain-tum.yaml"/>

  <!-- Image Processor Nodelet  包含了image.launch文件-->
  <include file="$(find msckf_vio)/launch/image_processor_livo.launch">
    <arg name="robot" value="$(arg robot)"/>
    <arg name="calibration_file" value="$(arg calibration_file)"/>
  </include>

  <!-- Msckf Vio Nodelet  -->
  <group ns="$(arg robot)">

   <!--这行配置指定了使用 nodelet 包的 nodelet 类型执行文件来启动一个名为 vio 的Nodelet实例。-->
    <node pkg="nodelet" type="nodelet" name="vio"
      args='standalone msckf_vio/MsckfVioNodelet'
      output="screen">

      <!-- Calibration parameters -->
      <rosparam command="load" file="$(arg calibration_file)"/>

      <param name="publish_tf" value="true"/>
      <param name="frame_rate" value="10"/>
      <param name="fixed_frame_id" value="$(arg fixed_frame_id)"/>
      <param name="child_frame_id" value="odom"/>
       <!-- modified 5->30 -->
      <param name="max_cam_state_size" value="5"/>
      <param name="position_std_threshold" value="8.0"/>

      <param name="rotation_threshold" value="0.2618"/>
      <param name="translation_threshold" value="0.1"/>
      <param name="tracking_rate_threshold" value="0.5"/>

      <!-- Feature optimization config -->
      <param name="feature/config/translation_threshold" value="-1.0"/>

      <!-- These values should be standard deviation -->

      <!-- modified:imu:handsfree -->
      <param name="noise/gyro" value="1.75746e-03"/>
      <param name="noise/acc" value="3.21412e-02"/>
      <param name="noise/gyro_bias" value="6.04908e-06"/>
      <param name="noise/acc_bias" value="6.02613e-04 "/>
      
      <param name="noise/feature" value="0.035"/>

      <param name="initial_state/velocity/x" value="0.0"/>
      <param name="initial_state/velocity/y" value="0.0"/>
      <param name="initial_state/velocity/z" value="0.0"/>

      <!-- These values should be covariance -->
      <param name="initial_covariance/velocity" value="0.25"/>
      <param name="initial_covariance/gyro_bias" value="0.01"/>
      <param name="initial_covariance/acc_bias" value="0.01"/>
      <param name="initial_covariance/extrinsic_rotation_cov" value="3.0462e-4"/>
      <param name="initial_covariance/extrinsic_translation_cov" value="2.5e-5"/>

      <!-- modified:imu:handsfree -->
      <remap from="~imu" to="/handsfree/imu"/>
      <remap from="~features" to="/firefly_sbx/image_processor/features"/>

    </node>
  </group>

  <!-- GroundTruth node -->

  <!-- Start RViz -->
  <node name="rviz" pkg="rviz" type="rviz" args="-d $(find msckf_vio)/rviz/rviz_euroc_config.rviz">
    <!-- Set the RViz configuration file if you have a specific setup -->
    <param name="use_sim_time" value="true"/>  <!-- Use simulation time -->
  </node>
</launch>
